### ISSUE TITLE
milkie: results with several categories

### ISSUE THREAD

**Issue by ngosang** [Jackett/Jackett#8365]
![image](https://user-images.githubusercontent.com/10577978/80292871-56b91d00-875a-11ea-9638-f72b43a92ab4.png)

@milkiecc I think @garfield69 can help you

**Comment by garfield69** [comment]
a search for `red town` returns the following json
```
{"hits":3,"took":10,"torrents":[{"id":"AWXCDGXr7PRK","releaseName":"Red_Town_RIP-DVN","category":4,"createdAt":"2020-04-26T00:04:21+02:00","size":388416930,"downloaded":3,"seeders":4,"partialSeeders":0,"leechers":0,"verified":false,"metainfo":{"root":"","files":null},"groupId":0,"externals":{"imdb":null,"tmdb":null,"tvmaze":null}},{"id":"VX1L3LUwx8wk","releaseName":"Red_Town_RIP-DVN","category":6,"createdAt":"2020-04-26T00:04:07+02:00","size":538416930,"downloaded":4,"seeders":5,"partialSeeders":0,"leechers":0,"verified":false,"metainfo":{"root":"","files":null},"groupId":0,"externals":{"imdb":null,"tmdb":null,"tvmaze":null}},{"id":"BMB1qXbedZ1o","releaseName":"Red.Town-DARKSiDERS","category":4,"createdAt":"2020-04-25T14:45:52+02:00","size":1302898009,"downloaded":3,"seeders":3,"partialSeeders":0,"leechers":0,"verified":false,"metainfo":{"root":"","files":null},"groupId":0,"externals":{"imdb":null,"tmdb":null,"tvmaze":null}}]}
```
the c# indexer for this site has the following categories defined:
https://github.com/Jackett/Jackett/blob/475e19441dee64d6437d9cae0769b1fa0b04776c/src/Jackett.Common/Indexers/Milkie.cs#L39-L52
and the following logic for returning a categories result:
https://github.com/Jackett/Jackett/blob/475e19441dee64d6437d9cae0769b1fa0b04776c/src/Jackett.Common/Indexers/Milkie.cs#L120

normally a site would have a unique category id for each cat.
But since your json does not have additional details to narrow down to a subcat, we don't have any way to pick the subcat we want, and we are going to have to go for a general category instead.
I will amend the indexer.

**Comment by garfield69** [comment]
Jackett 0.16.137

**Comment by milkiecc** [comment]
@ngosang @garfield69 Right, so we actually have "subcategories", but we don't give them unique IDs, instead it's read from some other metadata. Initially I did it your way too, but as far as I can understand now there's no way to actually search for console games for example, and `AddCategoryMapping("6", TorznabCatType.PC, "Apps")` has `PCGames` in it, so that would be overlapping with `AddCategoryMapping("4", TorznabCatType.PCGames, "Games");`

**Comment by cadatoiva** [comment]
@garfield69, Unfortunately, by deleting the duplicate lines, you also make it so that the search the other way `TorznabCapsToTrackers` will not map correctly. We can change the category logic on the results to only pick one, but we can't do this the other way around. We did talk about this a bit in #7761.

**Comment by cadatoiva** [comment]
Looking at it further, the `Result.Category` is an `ICollection` type, so it is intentional for a tracker to return multiple categories when a tracker result fits multiple categories. In this case, since we are unable to determine the sub-category the original issue shouldn't necessarily be considered a bug. Older trackers also make use of this feature, like how it was necessary to add back in a duplicate category mapping in https://github.com/Jackett/Jackett/pull/7761#discussion_r395969992 

And it can also be seen used in the way Milkie is using it in BB. https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Indexers/BB.cs#L50-L63

**Comment by cadatoiva** [comment]
List of C# trackers with duplicate mapping entries:
```
Find all "AddCategoryMapping\(([^,]+),[\w\W\n]+?AddCategoryMapping\(\1,", Regular expressions, Find Results Table, Current Project: Jackett.Common\Jackett.Common.csproj, ""
src\Jackett.Common\Indexers\Abstract\AvistazTracker.cs(48):            AddCategoryMapping(1, TorznabCatType.Movies);
src\Jackett.Common\Indexers\Abstract\AvistazTracker.cs(52):            AddCategoryMapping(2, TorznabCatType.TV);
src\Jackett.Common\Indexers\BB.cs(50):            AddCategoryMapping(1, TorznabCatType.Audio);
src\Jackett.Common\Indexers\BB.cs(59):            AddCategoryMapping(10, TorznabCatType.TVHD);
src\Jackett.Common\Indexers\DigitalHive.cs(56):            AddCategoryMapping(38, TorznabCatType.PCPhoneIOS);
src\Jackett.Common\Indexers\DigitalHive.cs(71):            AddCategoryMapping(17, TorznabCatType.ConsolePS3);
src\Jackett.Common\Indexers\DigitalHive.cs(81):            AddCategoryMapping(33, TorznabCatType.ConsoleWii);
src\Jackett.Common\Indexers\DigitalHive.cs(83):            AddCategoryMapping(45, TorznabCatType.ConsoleXbox);
src\Jackett.Common\Indexers\FileList.cs(61):            AddCategoryMapping(13, TorznabCatType.TVSport, "Sport");
src\Jackett.Common\Indexers\GimmePeers.cs(49):            AddCategoryMapping(11, TorznabCatType.ConsolePS3);
src\Jackett.Common\Indexers\GimmePeers.cs(52):            AddCategoryMapping(12, TorznabCatType.ConsoleXBOX360DLC);
src\Jackett.Common\Indexers\GimmePeers.cs(58):            AddCategoryMapping(21, TorznabCatType.TV);
src\Jackett.Common\Indexers\GimmePeers.cs(65):            AddCategoryMapping(50, TorznabCatType.XXX);
src\Jackett.Common\Indexers\Partis.cs(82):            AddCategoryMapping(8, TorznabCatType.AudioOther, "Glasba/Ostalo");
src\Jackett.Common\Indexers\PassThePopcorn.cs(52):            AddCategoryMapping(1, TorznabCatType.Movies, "Feature Film");
src\Jackett.Common\Indexers\Pretome.cs(114):            AddCategoryMapping(trackerCategory, newznabCategory);
src\Jackett.Common\Indexers\RevolutionTT.cs(86):            AddCategoryMapping("40", TorznabCatType.ConsoleWii);
src\Jackett.Common\Indexers\RevolutionTT.cs(147):            AddCategoryMapping("36", TorznabCatType.Books);
src\Jackett.Common\Indexers\toloka.cs(108):            AddCategoryMapping(137, TorznabCatType.MoviesDVD, "|-Мультфільми та мультсеріали в DVD");
```

List of YAML trackers that duplicate category IDs to search multiple Torznab categories:
```
Find all "id: (\d+),[\w\W\n]+?id: \1\D", Regular expressions, Find Results Table, Current Project: Jackett.Common\Jackett.Common.csproj, ""
src\Jackett.Common\Definitions\Bittorrentfiles.yml(80):      - {id: 35, cat: Books/Ebook, desc: "Sonstiges e-Book-DE"}
src\Jackett.Common\Definitions\cinemamovies.yml(20):      - {id: 2, cat: TV/Anime, desc: "Anime"}
src\Jackett.Common\Definitions\gay-torrentsorg.yml(53):      - {id: 68, cat: XXX, desc: "Homo Erotic"}
src\Jackett.Common\Definitions\gay-torrentsorg.yml(70):      - {id: 41, cat: Movies, desc: "Non-Porn"}
src\Jackett.Common\Definitions\korsar.yml(61):      - {id: 109, cat: Movies, desc: "Бенефис. Мастера искусств отечественного Театра и Кино"}
src\Jackett.Common\Definitions\speedtorrentreloaded.yml(34):      - {id: 84, cat: Console/PS3, desc: "PS3/PS4"}
src\Jackett.Common\Definitions\torrentbd.yml(32):      - {id: 14, cat: Console/Xbox, desc: "Games - Xbox"}
src\Jackett.Common\Definitions\trezzor.yml(19):      - {id: 14, cat: PC/Games, desc: "Hry"}
src\Jackett.Common\Definitions\trezzor.yml(22):      - {id: 18, cat: Books, desc: "Knihy CZ/SK lokalizace"}
src\Jackett.Common\Definitions\world-of-tomorrow.yml(34):      - {id: 84, cat: Console/PS3, desc: "PS3/PS4"}
```

@garfield69 @ngosang how do these trackers behave when you do a search in one of the categories?

**Comment by garfield69** [comment]
If the milkie json does not provide enough data points to uniquely identify a subcat (either via a specific subcat id or tags containing the subcat label which could be used by the code to case-switch a particular cat) then by using duplicate cats with different descriptions (as the code originally did) does not really provide a method for filtering searches that return the intended results.
If a user filters with console-cat-4 they should expect to see only console games returned, but they are going to get all the pcgames-cat-4 as well, which was not what the user intended. Far better in this instance provide only a filter for games-cat-4 and at least this way the results line up with the intent.

**Comment by cadatoiva** [comment]
@garfield69, then all the above categories should be changed as well, because each one of those is doing the same thing.

**Comment by milkiecc** [comment]
I can return more metadata in the json response that would help identify the exact category. It would be a big switch case though, which is what I wanted to avoid in the first place.

**Comment by garfield69** [comment]
@cadatoiva  how did this escalate into a hunt for a bug?
I have no idea what the original coders intended when they set up the catmaps, and how other indexers ended up with duplicate categories is probably the result a best fit to a messy miss-match between data points and filtering.

I generally approach this from the point of view of a user.
And if the user turns on a filter then the filter should try to return only results as described by that filter label.
If the label is too specific and the filter returns results that do not fit then IMO
- either the label should be made more general to encompass what is actually returned
- or if there are data points available that can help narrow down the selection of the category then the code should attempt to do it, so that the filter works as the label describes

in the case of milkie, cat6 originally described multiple software platforms and an attempt to filter by any one of the software subcats returns all those subcats since we only have one data point it its not enough to identify the subcat. 
So I prefer to see one general cat called software than a list of subcats that cannot be used to filter each subcat specifically.

**Comment by cadatoiva** [comment]
@garfield69, it's not that it's a hunt for a bug per se, just that there needs to be a discussion on what the best way to implement this. As seen from above there are several trackers that implement broad tracker categories this way.

From a user standpoint, if I'm trying to find a xbox360 game on Milkie, I'd have to either search all categories, or know that PC/Games includes console games. I can't use the Console or Console/Xbox360 cat to get any results from this tracker now. Yes, it does mean on the results side we are also returning PC games, PS3/4, Xbox One, and Wii games with our Xbox 360 results.

Basically from a user side, they're going to be confused either way, but I think it'd be preferred to have too many results we can filter, instead of inexplicably returning 0 results because they picked the filter we chose to omit.

**Comment by garfield69** [comment]
![image](https://user-images.githubusercontent.com/6933240/80319170-23ae8080-8863-11ea-90b2-8b9ffc230f85.png)
Yes, things are not clear right now for the user. The filter list shows both the torznab 4050 (PC/Games) and the custom cat 100004 (Games)
And while both filters return the same results (both console and pcgames), only the 100004 indicates the correct intent of the filter in this case.
One option might be, that it there are custom categories then the the jackett dashboard search panel should not display the torznab cats in the filter drop down.

**Comment by cadatoiva** [comment]
That works when searching individual trackers, but wouldn't work when doing any kind of aggregate search, which currently disables indexer custom categories. At that point, we end up with the same issue of either not getting results on a category when we should (by omitting the duplicate categories) or getting extra results not relevant to the filtered category (by including the duplicated categories).

We have features to filter results after they're returned in the UI, and Sonarr/Radarr/etc have search behavior that encourages extra/irrelevant results (Sonarr sends a full season search when automatically searching for an episode) so they can do their own filtering after results are returned. I think it would be better if we defaulted to include more results (even if potentially irrelevant) instead of removing categories of results just because they don't fit neatly into the Torznab category boxes.
